### ISSUE TITLE
Dismax Boost may not be set correctly for weighting fields in queries

### ISSUE THREAD

**Issue by mistcat** [pantheon-systems/solr-power#371]
__in includes/class-solrpower-api.php__

the plugin currently calls:

```php
$solr_boost_query = apply_filters( 'solr_boost_query', 'post_title^2 post_content^1.2' );
			if ( false !== $solr_boost_query ) {
				$dismax->setBoostFunctions( $solr_boost_query );
			}
```
However I believe what is should call instead in order to correctly weight the fields as described is:

```php
$solr_boost_query = apply_filters( 'solr_boost_query', 'post_title^2 post_content^1.2' );
			if ( false !== $solr_boost_query ) {
				$dismax->setQueryFields( $solr_boost_query );
			}
```

**Comment by danielbachhuber** [comment]
Thanks for the report, @mistcat.

> However I believe what is should call instead in order to correctly weight the fields as described is:

Can you provide more background information for this suggestion?

For context, `$dismax->setBoostFunctions` was originally added in bfcc9fd8de9bc099e838a4e619acdd010c8cb7ff and became filterable with 003195d25622254bfc3e139154225ffddcf128e5

**Comment by danielbachhuber** [comment]
> I'd be happy to submit a PR, but 1 is a bit of a blocker for me as I really don't understand why the switch to `q.alt` was made in #143. I'd also appreciate it if someone with more Solr experience weighed in here.

I imagine it was a "tried something else and it worked" implementation. Because it's only used for fetching stats, we could run an entirely separate Solr query for it, and ignore the code path used for the main Solr query.

I'd be fine with a pull request including your suggested changes as long as they also include a reasonable amount of test coverage around them.

**Comment by mgburns** [comment]
Looks like this is a bit more involved than I originally thought. The reason that you're using `q.alt` instead of `q` because of the way all the filter queries are set up in [`SolrPower_WP_Query::build_query()`](https://github.com/pantheon-systems/solr-power/blob/master/includes/class-solrpower-wp-query.php#L440-L510). For example, here's the query it builds to filter by post type:

  ```
  star wars AND(post_type:post)
  ```

The dismax parser, being geared towards humans, doesn't support that syntax for filtering. The standard query parser, which is used for `q.alt`, does and is therefore required.

I think the bigger question is whether we should use the standard query parser or dismax by default. My feeling is dismax, for all the reasons described in the [docs overview](https://lucene.apache.org/solr/guide/6_6/the-dismax-query-parser.html). It's less powerful, but 99% of search users won't know the difference. It's also more forgiving / less error prone, which feels appropriate as well.

I don't currently have the bandwidth to PR a switch in default parser, but I imagine the approach would be switching to [filter queries](https://solarium.readthedocs.io/en/latest/queries/select-query/building-a-select-query/adding-filterqueries/) for post type / status / tax / meta filters (which I believe will also be more performant).

The one thing I do recommend is removing the `$dismax->setBoostFunctions()` call / filter that started this issue. It's not doing what it says it's doing, and completely throws off relevancy for sites with more than a handful of posts. (The unit tests don't have enough posts to demonstrate this).

**Comment by defunctl** [comment]
Can confirm that modifying `solr_boost_query` with all kinds of different variations never changes the search results.

**Comment by jimkeller** [comment]
I don't think we should be using the q.alt - after a lot of debugging it doesn't seem to support the boost syntax. 

I think using q.alt was a misunderstanding of the note under "selecting all documents" at the following link - it only applies if you're trying to use standard solr syntax in a dismax query: https://solarium.readthedocs.io/en/stable/queries/select-query/building-a-select-query/components/dismax-component/

To get boosting working, I had to do the following:

Comment out the part in class-solrport-api.php that uses q.alt:

`	function dismax_query( $query, $dismax ) {
		// $dismax->setQueryAlternative( $query->getQuery() );
		// $query->setQuery( '' );

		return $query;
	}`

Get rid of the default boostFunctions because it's the incorrect way to set them (this is in a custom plugin):
`add_filter( 'solr_boost_query', function ( $solr_boost_query ) {
  return false; 
 }, 10, 1 );`

Add them in manually:
`add_filter( 'solr_dismax_query', function( $dismax, $query ) {
  $dismax->setQueryFields('text^1.5 post_title^4 boosted_keywords^8');
}, 10, 2 );`

**Comment by mrobit** [comment]
@jimkeller I have no idea how any of this works, but this made our search queries work a lot more reliably. Kudos.

**Comment by jimkeller** [comment]
I created a pull request at: https://github.com/pantheon-systems/solr-power/pull/495 to workaround this problem.
The pull request allows you to define a constant SOLRPOWER_DISABLE_QUERY_ALT that will circumvent the code that causes the boost not to work. 

So simply put the following in wp-config.php:
`define('SOLRPOWER_DISABLE_QUERY_ALT', true);`

Note that you will still have to set your own boost filter as shown:
`add_filter( 'solr_dismax_query', 
  function( $dismax, $query ) {
    $dismax->setQueryFields('text^1.2 post_title^2 boosted_keywords^5');
  }
, 10, 2 );`

Also attached is a patch against the master branch. 
[wp-solrpower-disable-query-alt.patch.txt](https://github.com/pantheon-systems/solr-power/files/6418481/wp-solrpower-disable-query-alt.patch.txt)

**Comment by danielbachhuber** [comment]
Shipped the constant with v2.2.4. Thanks again for your pull request, @jimkeller !

**Comment by mistcat** [comment]
So exciting to see this get sorted out!  Great work @jimkeller & @mgburns !  I'm sorry I didn't have more time to help contribute a fix myself, but am glad to see all the lovely community support.  Let me know if I can help roadtest anything or contribute to documentation for it!
